Resolved 2. Feature: Implement a read status for messages

Ed1196 opened the pull request:
This implements tracking of read messages by the recipient. 
Both client and server programs had to be changed. 

- On the server side I added an extra read boolean field to the Messages model. It will be default be false, and marked true once a user has read it. A new endpoint was added to messenger_backend called 'update-messages' that will change all messages read boolean field to True. 
- On the client side clicking on a chat window will call 'update-messages' to marked whatever unread messages we have in that chat to true. 

First user **test2** sends two messages to **test1**
![image](https://user-images.githubusercontent.com/34897031/139939383-a0c160e3-f7c9-475d-ba6d-6209e884cce7.png)

User **test1** is not in the active chat, so a badge is used to notify him of **test2's** unread messages
![image](https://user-images.githubusercontent.com/34897031/139940011-9a33ea3d-bd4d-4484-88a6-0e77196f8bf2.png)

When we click on the active chat for user **test2** those messages will be marked as read for user **test1** and the badge disappears 
![image](https://user-images.githubusercontent.com/34897031/139940236-ea326e8a-1676-49b1-8be8-e2ff3ba4f69c.png)

Finally we reply back
![image](https://user-images.githubusercontent.com/34897031/139940289-669f2738-f186-472e-a318-af5ceaf29d93.png)

On page reload, those messages are not longer reflected as a badge notification
![image](https://user-images.githubusercontent.com/34897031/139941150-62dd593d-f78d-42b5-80c4-b17e8a22dc41.png)

Ed1196 committed:
1. Bug fix: Sending messages was fixed, problem with how state was updated in reducer functions

Ed1196 committed:
Implemented a read status for messages

Ed1196 committed:
Removed console.log's

Ed1196 committed:
Added a read field to Messages and create a new endpoint to update messages

Ed1196 committed:
Changed the way messages are returned, I reverse the list before returning it. lastMessageeText was also changed to point to the end of the list

Ed1196 committed:
Removed reversal of messages in Messages component, changed how messages are added to state

Ed1196 committed:
Created PR request

Ed1196 committed:
Solved conflicts for conversations.py

Ed1196 committed:
Submit for PR review

Ed1196 committed:
Removed some console.log's

Ed1196 committed:
Removed .idea foldeer

Ed1196 committed:
Removed .idea and Pipfile from server files and added them to gitignore

johnsciutto commented on client/src/components/ActiveChat/Input.js:
```diff
@@ -32,8 +31,8 @@ const Input = (props) => {
       recipientId: otherUser.id,
-      conversationId,
-      sender: conversationId ? null : user
+      conversationId: conversationId,
```
You actually don't need to do this, since in JS, if the key of the property shares the same name as the variable holding the value, then it can be simplified:

```js
// this...
const obj = {
  specialProp: specialProp
}

// becomes...
const obj = { specialProp }
```

Ed1196 replied:
These were removed

johnsciutto commented on client/src/components/ActiveChat/Input.js:
```diff
@@ -20,8 +20,7 @@ const useStyles = makeStyles(() => ({
   const [text, setText] = useState("");
-  const { postMessage, otherUser, conversationId, user } = props;
-
+  const { postMessage, otherUser, conversationId, user, activeConversation } = props;
```
I don't see you using `activeConversation` here, so you can delete it 🙂

Ed1196 replied:
activeConversation was removed from props

johnsciutto commented on client/src/components/ActiveChat/Input.js:
```diff
@@ -55,6 +54,12 @@ const Input = (props) => {
+    activeConversation: state.activeConversation,
+  };
+}
+
```
Same as above :slightly_smiling_face:

Ed1196 replied:
mapStateToProps was removed

johnsciutto commented on client/src/components/Sidebar/Chat.js:
```diff
@@ -23,11 +23,31 @@ const Chat = (props) => {
   const { otherUser } = conversation;
+  const { messages } = conversation;
+  // Will disable badge showing unread messages when we are currently in the chat window
+  const { activeConversation } = props
```
You can group these destructurings:

```suggestion
  const { conversation, activeConversation } = props;
  const { otherUser, messages } = conversation;
```

Ed1196 replied:
destructuring was done in groups

johnsciutto commented on client/src/components/Sidebar/Chat.js:
```diff
@@ -23,11 +23,31 @@ const Chat = (props) => {
+      }
+    };
+    return accumulator;
+  }, 0);
```
While this works, it's an operation that could be slow depending on the number of messages in the conversation. It might be a better idea to calculate this number outside of the component (where it is re-run on every re-render of this component). It might even be a good idea to use the [`count`](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#django.db.models.query.QuerySet.count) method in Django to let the query do this calculation in the back-end 🙂

Ed1196 replied:
All this logic was moved to the server side

johnsciutto commented on client/src/components/Sidebar/Sidebar.js:
```diff
@@ -3,45 +3,78 @@ import { Box, Typography } from "@material-ui/core";
+      updatedConvos[convoIndx].messages = newConvo.convo.messages;
+      gotConversations(updatedConvos); 
+    }
+  };
```
While this works, it also limits you to mark the messages as read when you click on the chat. But if the chat is already open and new messages come in, then the messages are not marked as read, and the user has to click on the chat sidebar again to mark them as read 🤔

Ed1196 replied:
Messages are marked as read if two users are in the same active chat and are sending messages to each

johnsciutto commented on client/src/components/Sidebar/Sidebar.js:
```diff
@@ -3,45 +3,78 @@ import { Box, Typography } from "@material-ui/core";
+   When a user click on a chat box, the list of unread messages will be marked read on the db.
+   Once marked read, the new form of that conversation is returned to use and we can update that one conversations
+   in the conversations state. 
+   */
```
You can delete these types of explanatory comments 🙂 Your code so far is easy to read, so it speaks for itself 👍🏻

Ed1196 replied:
This comment was removed

johnsciutto commented on client/src/components/Sidebar/Sidebar.js:
```diff
@@ -3,45 +3,78 @@ import { Box, Typography } from "@material-ui/core";
-          return <Chat conversation={conversation} key={conversation.otherUser.username} />;
+          return (
+            <Chat
+              user = {props.user}
```
```suggestion
              user={props.user}
```

Ed1196 replied:
These changes were made

johnsciutto commented on client/src/components/Sidebar/Chat.js:
```diff
@@ -23,11 +23,31 @@ const Chat = (props) => {
+  of unread messages that we have. 
+  */
+  let unreadCount = messages.reduce((accumulator, currValue) => {
+    if(currValue.senderId !== props.user.id) {
```
I see you're getting the `user` here by doing prop-drilling, but you have access to it through Redux, so it would be better to pass it in in the `mapStateToProps`, and if you're only using the id, you can just pass that 🙂

Ed1196 replied:
This logic was completely removed and user is no longer passed as a prop

johnsciutto commented on client/src/store/utils/thunkCreators.js:
```diff
@@ -109,6 +108,15 @@ export const postMessage = (body) => (dispatch) => {
 
+export const updateMessages = async (body) => {
+  try {
+    const { data } = await axios.post("/api/update-messages", body);
```
Nice work here!
But there's a couple of things I wanted to mention about the API design:

- Because this route is meant to update a resource instead of creating one, the better verb to use would be `PATCH` or `PUT` instead of `POST`.
- It's better practice to have the name of the endpoint be a noun (usually the name of the resource being updated, created or retrieved) and not a verb. So in this case something like "/api/messages/read-status" might be an option.

Ed1196 replied:
Both of these were changed

johnsciutto commented on server/messenger_backend/models/message.py:
```diff
@@ -7,6 +7,7 @@
 class Message(utils.CustomModel):
     text = models.TextField(null=False)
     senderId = models.IntegerField(null=False)
+    read = models.BooleanField(null=False, default=False)
```
💯

johnsciutto commented on server/messenger_backend/views/api/updatemessages.py:
```diff
@@ -0,0 +1,38 @@
+        try:
+            user = get_user(request)
+
+            if user.is_anonymous:
+                return HttpResponse(status=401)
+
```
💯 
We also want to protect this route if the person calling the route isn't part of the conversation (with a `403` forbidden)

Ed1196 replied:
A 403 status error is now returned when the user asking for the request is not part of the conversation

johnsciutto commented on server/messenger_backend/views/api/updatemessages.py:
```diff
@@ -0,0 +1,38 @@
+                for m in messages:
+                    if m.senderId != user.id:
+                        m.read = True
+                        m.save()
```
I don't love the idea of doing this in a loop, it might be more efficient from a database perspective to make a batch update using a filter and update methods. [Maybe something like these examples](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#django.db.models.query.QuerySet.update)

Ed1196 replied:
Update was change from loop to batch update

johnsciutto reviewed (commented):
Hi @Ed1196,
Nice work on this feature so far 🙂 .
I see some changes that are from the previous PR. When you create the PR for the last issue, make sure that you're branching off of the `main` branch, and not from the branch used for this issue, otherwise your PR will contain the changes made in the previous issue as well.
I also noticed a few things from the implementation that are missing (which are present in the given design mocks):

- There is an avatar of the other user under their last read message.
- If the last message is unread, the preview text in the sidebar is bolded.

(See the mocks below for reference)
![PI - Missing Text and Avatar](https://user-images.githubusercontent.com/31010488/140077928-5b10989d-9273-4f21-a1c1-b84713fe738f.jpg)

Once the changes have been implemented, lets us know on Slack using the "/hatchy pr" command again and another reviewer will come come by to give approval 🙂

Ed1196 committed:
Solved the following issues
- Last unread message is bolded
- Removed redundant key properties
- Removed unused props
- Grouped destructurings
- Moved count of unread messages to the server
- Handle case where messages are marked as read when chat is already open

Ed1196 committed:
Implemented the following
- New unread messages are bolded
- User avatar is now shown on last seen message
- Added extra fields for convo and messages when returned from server

Ed1196 committed:
Implemented real time read messages that display a user's avatar on the read message

Ed1196 committed:
Solved bug where last read message avatar not showing up, when clicking on a chat with unread messages

Ed1196 committed:
Did the following:
- Fixed bugs with real time update of placing avatar on last read message
- Implemented one more socket event
- Changed how messeges are updated when a user clicks on a chat box

Ed1196 committed:
Changed update message resource from post to put
Changed name of the endpoint to be a noun

Ed1196 committed:
Added protection to updatemessages route which returns a 403 if the user who requested it is not part of the conversation

Ed1196 committed:
Changed update of unread messages from using a loop to using a batch update using exclude and update

Ed1196 committed:
Fixed the bug that created a duplicate of a convo when
two users were on the active chat window and on of them sends a message for the first time.
Because of the way the state of conversations was being done, a new convo was added on top
of already existing one. Now that convo is replaced.

Ed1196 committed:
Solved bugs for the following scenarios
- Showing a new unread message for a new convo on the receiver user when he has nothing typed on the Search chats bar
- Showing a new unread message for a new convo on the receiver user when he has something typed on the Search chats bar
The scenarios arise when we are creating new conversations, but the search feature causes them to either create a duplicate or
not showing new incoming messages.

Ed1196 committed:
Removed console.log's

Ed1196 commented:
Updated UI:

User **test1** sends a message to user **test2** who is currenlty not in the active chat window for this conversation
<img width="640" alt="Screen Shot 2021-11-05 at 5 11 16 PM" src="https://user-images.githubusercontent.com/34897031/140579826-44e7e05f-692d-41c7-b352-caa9c1549383.png">

User **test2** gets notified of the new messages which are displayed by a badge and a bolded latest message
<img width="304" alt="Screen Shot 2021-11-05 at 5 11 58 PM" src="https://user-images.githubusercontent.com/34897031/140580117-9f4b8629-17ff-4917-8d60-19d9275a009b.png">

Once user **test2** clicks on the chat box, the notifications disappear
<img width="746" alt="Screen Shot 2021-11-05 at 5 12 07 PM" src="https://user-images.githubusercontent.com/34897031/140580147-e0e88052-446c-4dd1-87a0-8d239e2a4c5e.png">

Once user **test2** clicked on the chat box, **test1** gets notified that those messages were read.
The last message read by user **test2** for **test1** is reflected by an avatar under the last read message.
<img width="640" alt="Screen Shot 2021-11-05 at 5 12 13 PM" src="https://user-images.githubusercontent.com/34897031/140580496-3818858e-143c-4ccf-9cfb-eed221524fcd.png">

User **test2** sends a message to user **test1** who left the active window after the **"and you?** message, so the avatar under the
last message does not move. 
<img width="738" alt="Screen Shot 2021-11-05 at 5 14 07 PM" src="https://user-images.githubusercontent.com/34897031/140581207-db2c42c3-3f33-4d36-9815-cba3261a3b91.png">

andreluciani commented on client/src/components/ActiveChat/Messages.js:
```diff
@@ -11,9 +11,21 @@ const Messages = (props) => {
+            text={message.text}
+            time={time}
+            otherUser={otherUser}
+            id={message.id}
```
The `id` is already passed as `key` prop, no need to pass it again.

andreluciani commented on client/src/components/ActiveChat/SenderBubble.js:
```diff
@@ -1,41 +1,54 @@
       <Box className={classes.bubble}>
         <Typography className={classes.text}>{text}</Typography>
       </Box>
+      {lastReadByOtherId === id && (
+        <Avatar
+          alt={props.otherUser.username}
+          src={props.otherUser.photoUrl}
+          className={classes.avatar}
+        ></Avatar>
+      )}
```
:100:

andreluciani commented on client/src/components/Sidebar/ChatContent.js:
```diff
@@ -15,17 +15,21 @@ const useStyles = makeStyles((theme) => ({
+    letterSpacing: (props) =>
+      props.unread > 0 && props.disableUnread ? "0.3px" : "-0.17",
+    textShadow: (props) =>
+      props.unread > 0 && props.disableUnread ? ".5px 0 black" : "",
```
The same ternary operator is used several times, can you refactor this to reduce repetition?

andreluciani commented on client/src/socket.js:
```diff
@@ -18,8 +20,28 @@ socket.on("connect", () => {
+          conversationId: data.message.conversationId,
+        });
+      }
+    }
```
I suggest using reducers and/or thunks to implement the logic and dispatching them here in order to keep the code organized and easier to maintain.

andreluciani commented on client/src/store/utils/thunkCreators.js:
```diff
@@ -110,7 +112,7 @@ export const postMessage = (body) => async (dispatch) => {
 export const updateMessages = async (body) => {
   try {
-    const { data } = await axios.post("/api/update-messages", body);
+    const { data } = await axios.put("/api/messages/read-status", body);
```
:100:

andreluciani commented on server/messenger_backend/views/api/conversations.py:
```diff
@@ -40,12 +40,15 @@ def get(self, request: Request):
                         message.to_dict(["id", "text", "senderId", "createdAt", "read"])
                         for message in convo.messages.all().reverse()
                     ],
+                    "unread": convo.messages.filter(read=False).exclude(senderId=user_id).count(),
```
:100:

andreluciani commented on server/messenger_backend/views/api/updatemessages.py:
```diff
@@ -8,31 +8,51 @@
-                    if m.senderId != user.id:
-                        m.read = True
-                        m.save()
+                convo = Conversation.objects.filter(id=conversation_id).first()
+                if user_id not in (convo.user1.id, convo.user2.id):
+                    return HttpResponse(status=403)
```
:100:

andreluciani reviewed (approved):
Nice work addressing John's comments!
I've left some additional comments on the code, also I've noticed a small problem with your implementation:
- When receiving messages on a chat that is active, it works properly, but if the user changes to another chat, it shows the unread badge, as if the last messages received were not read.

PR is approved! Feel free to merge it after addressing the comments :slightly_smiling_face:

Ed1196 committed:
Fixed bug of messages not being marked as read if a user receives a message while
in the active chat windo and switches to another active chat

Ed1196 committed:
ternary operator used in ChatContent is compute before being sent to useStyles to avoid repetition

Ed1196 committed:
Moved most of the logic from ChatContent to reducerFunctions. The logic is now dispatched inside the
setNewUnread reducer function

Ed1196 committed:
All bugs have been fixed. Branch is ready to be merged
